### PR TITLE
Bmatsuo/reduce allocations

### PR DESCRIPTION
**Pull request by bmatsuo-at-luthersystems**
The diff looks a bit all over the place. But, looking at the individual commits would give a more focused view of the changes.

### CODE REVIEW

**Comment by bmatsuo-at-luthersystems** [comment]
These are the benchmark improvements that I used benchstat to format.

```
name                       old time/op    new time/op    delta
pkg:github.com/luthersystems/elps/lisp goos:darwin goarch:amd64
EnvGet-4                     55.5ms ± 0%    18.7ms ± 0%  -66.29%
EnvFunCallBuiltin-4          2.00ms ± 0%    1.13ms ± 0%  -43.39%
EnvFunCallRecursion-4        20.2ms ± 0%    11.2ms ± 0%  -44.69%
EnvFunCallTailRec-4          12.0ms ± 0%     7.4ms ± 0%  -37.84%
EnvFunCallPos-4              5.63ms ± 0%    3.27ms ± 0%  -41.93%
EnvFunCallKey-4              5.94ms ± 0%    3.77ms ± 0%  -36.49%
EnvFunCallOpt-4              5.56ms ± 0%    3.05ms ± 0%  -45.18%
EnvFunCallVar-4              6.20ms ± 0%    3.57ms ± 0%  -42.39%
MacroDefun-4                 71.7µs ± 0%    50.7µs ± 0%  -29.30%
pkg:github.com/luthersystems/elps/lisp/lisplib/libjson goos:darwin goarch:amd64
Package/$load-4              0.00ns ± 0%    0.00ns ± 0%   -7.51%
Package/load-object-4        4.36ms ± 0%    3.51ms ± 0%  -19.49%
Package/load-array-4         4.35ms ± 0%    3.78ms ± 0%  -13.22%
Package/load-nested-4        6.55ms ± 0%    5.59ms ± 0%  -14.61%
Package/dump-object-4        5.17ms ± 0%    4.21ms ± 0%  -18.65%
Package/dump-array-4         3.45ms ± 0%    2.49ms ± 0%  -27.73%
Package/dump-nested-4        8.05ms ± 0%    7.27ms ± 0%   -9.77%
pkg:github.com/luthersystems/elps/parser/rdparser goos:darwin goarch:amd64
Parser/approx.lisp-4          239µs ± 0%     218µs ± 0%   -8.48%
Parser/complex.lisp-4         645µs ± 0%     605µs ± 0%   -6.29%
Parser/diff.lisp-4            271µs ± 0%     259µs ± 0%   -4.29%
Parser/scheme-math.lisp-4    1.43ms ± 0%    1.42ms ± 0%   -0.21%
Parser/sicp.lisp-4           1.03ms ± 0%    1.00ms ± 0%   -2.87%
Parser/stream.lisp-4          899µs ± 0%     863µs ± 0%   -4.06%
pkg:github.com/luthersystems/elps/parser/regexparser goos:darwin goarch:amd64
Parser/approx.lisp-4         1.65ms ± 0%    1.69ms ± 0%   +2.36%
Parser/complex.lisp-4        4.65ms ± 0%    4.60ms ± 0%   -1.10%
Parser/diff.lisp-4           2.09ms ± 0%    2.15ms ± 0%   +2.64%
Parser/scheme-math.lisp-4    12.0ms ± 0%    12.1ms ± 0%   +0.75%
Parser/sicp.lisp-4           7.92ms ± 0%    8.24ms ± 0%   +4.02%
Parser/stream.lisp-4         6.80ms ± 0%    6.80ms ± 0%   -0.01%

name                       old speed      new speed      delta
pkg:github.com/luthersystems/elps/parser/rdparser goos:darwin goarch:amd64
Parser/approx.lisp-4       7.21MB/s ± 0%  7.88MB/s ± 0%   +9.29%
Parser/complex.lisp-4      8.10MB/s ± 0%  8.64MB/s ± 0%   +6.67%
Parser/diff.lisp-4         6.57MB/s ± 0%  6.86MB/s ± 0%   +4.41%
Parser/scheme-math.lisp-4  8.16MB/s ± 0%  8.18MB/s ± 0%   +0.25%
Parser/sicp.lisp-4         6.89MB/s ± 0%  7.09MB/s ± 0%   +2.90%
Parser/stream.lisp-4       7.93MB/s ± 0%  8.27MB/s ± 0%   +4.29%
pkg:github.com/luthersystems/elps/parser/regexparser goos:darwin goarch:amd64
Parser/approx.lisp-4       1.04MB/s ± 0%  1.02MB/s ± 0%   -1.92%
Parser/complex.lisp-4      1.12MB/s ± 0%  1.14MB/s ± 0%   +1.79%
Parser/diff.lisp-4          850kB/s ± 0%   830kB/s ± 0%   -2.35%
Parser/scheme-math.lisp-4   970kB/s ± 0%   960kB/s ± 0%   -1.03%
Parser/sicp.lisp-4          890kB/s ± 0%   860kB/s ± 0%   -3.37%
Parser/stream.lisp-4       1.05MB/s ± 0%  1.05MB/s ± 0%    0.00%
```

**Review by sam-at-luther — APPROVED** [review]

File: lisp/builtins.go [hunk 1]
```diff
@@ -1489,12 +1489,63 @@ func builtinSlice(env *LEnv, args *LVal) *LVal {
 	if i > j {
 		return env.Errorf("end before start")
 	}
+
+	// Create an intermediate sliced list with a similar type
```
Comment (sam-at-luther) [inline lisp/builtins.go:1493]: is this the source of the bulk of the performance gains?

File: lisp/lisp.go [hunk 2]
```diff
@@ -1029,9 +1041,30 @@ func seqCells(v *LVal) []*LVal {
...
 	}
 }
+
+var defaultSourceLocation = &token.Location{
+	File: "<native code>",
+	Pos:  -1,
+}
+
+// TODO(elps2): make the LVal.Source "immutable" (possibly an interface or a
+// string) so it won't matter that nativeSource returns a shared reference.
```
[the remark this answers is not in the document]
Comment (bmatsuo-at-luthersystems) [reply lisp/lisp.go:1067]: Eh. I'm not too eager to create v2 tags given how go.mod works. These comments are intended to just be possible changes if we ever decide we need to create elps2.

File: lisp/builtins.go (hunk 1 above)
Comment (bmatsuo-at-luthersystems) [reply lisp/builtins.go:1493]: Ah. No. I don't have any benchmarks for the slice function in particular. And I don't think existing calls to slice would have their performance changed much. This intermediate slice is more to simplify additional type conversions that are required to support the new `'string` and `'bytes` type specifiers.

Comment (bmatsuo-at-luthersystems) [reply lisp/builtins.go:1493]: The performance improvements are primarily in lisp.go and env.go

**Pull request merged**